patch/llpcPatchDescriptorLoad.cpp

```diff
@@ -78,6 +78,85 @@ PatchDescriptorLoad::PatchDescriptorLoad()
     initializePatchDescriptorLoadPass(*PassRegistry::getPassRegistry());
 }
 
+// =====================================================================================================================
+/// Create a non-strided, non-swizzled buffer descriptor.
+///
+/// @param[in] pBaseAddr (i64) base address of the descriptor (top 16 bits are ignored);
+///                      for convenience, an <N x i32> is also accepted (using the low 48 bits)
+/// @param[in] pBufSize (i32) size of the referenced buffer in bytes
+/// @param[in] pInsertPoint instructions are inserted before pInsertPoint
+Value* PatchDescriptorLoad::buildBufferDescriptor(Value* pBaseAddr,
+                                                  Value* pBufSize,
+                                                  Instruction* pInsertPoint)
+{
+    IRBuilder<> builder(pInsertPoint);
+    Value* pBufDesc = UndefValue::get(m_pContext->Int32x4Ty());
+
+    Value* pElem0;
+    Value* pElem1;
+
+    if (pBaseAddr->getType() == m_pContext->Int64Ty()) {
+        pElem0 = builder.CreateTrunc(pBaseAddr, m_pContext->Int32Ty());
+        pElem1 = builder.CreateLShr(pBaseAddr, 32, "", true);
+        pElem1 = builder.CreateTrunc(pElem1, m_pContext->Int32Ty());
+    } else {
+        pElem0 = builder.CreateExtractElement(pBaseAddr, (uint64_t)0);
+        pElem1 = builder.CreateExtractElement(pBaseAddr, (uint64_t)1);
+    }
+
+    // DWORD0
+    pBufDesc = builder.CreateInsertElement(pBufDesc, pElem0, (uint64_t)0);
+
+    // DWORD1
+    pElem1 = builder.CreateAnd(pElem1, builder.getInt32(UINT16_MAX));
+    pBufDesc = builder.CreateInsertElement(pBufDesc, pElem1, (uint64_t)1);
+
+    // DWORD2
+    pBufDesc = builder.CreateInsertElement(pBufDesc, pBufSize, (uint64_t)2);
+
+    // DWORD3
+#if LLPC_BUILD_GFX10
+    const GfxIpVersion gfxIp = m_pContext->GetGfxIpVersion();
+    if (gfxIp.major < 10)
+#endif
+    {
+        SqBufRsrcWord3 sqBufRsrcWord3 = {};
+        sqBufRsrcWord3.bits.DST_SEL_X = BUF_DST_SEL_X;
+        sqBufRsrcWord3.bits.DST_SEL_Y = BUF_DST_SEL_Y;
+        sqBufRsrcWord3.bits.DST_SEL_Z = BUF_DST_SEL_Z;
+        sqBufRsrcWord3.bits.DST_SEL_W = BUF_DST_SEL_W;
+        sqBufRsrcWord3.gfx6.NUM_FORMAT = BUF_NUM_FORMAT_UINT;
+        sqBufRsrcWord3.gfx6.DATA_FORMAT = BUF_DATA_FORMAT_32;
+        LLPC_ASSERT(sqBufRsrcWord3.u32All == 0x24FAC);
+
+        pBufDesc = builder.CreateInsertElement(pBufDesc, builder.getInt32(sqBufRsrcWord3.u32All),
```

Comment:
You can see if we could write builder.CreateInsertElement(...) in one line. We allow 120 character.